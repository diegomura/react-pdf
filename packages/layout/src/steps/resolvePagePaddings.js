import { evolve, matchPercent } from '@react-pdf/fns';

/*
 * Translates page percentage horizontal paddings in fixed ones
 *
 * @param {Object} page container
 * @param {String} padding value
 * @returns {Object} translated padding value
 */
const resolvePageHorizontalPadding = container => value => {
  const match = matchPercent(value);
  return match ? match.percent * container.width : value;
};

/**
 * Translates page percentage vertical paddings in fixed ones
 *
 * @param {Object} page container
 * @param {String} padding value
 * @returns {Object} translated padding value
 */
const resolvePageVerticalPadding = container => value => {
  const match = matchPercent(value);
  return match ? match.percent * container.height : value;
};

/**
 * Translates page percentage paddings in fixed ones
 *
 * @param {Object} page
 * @returns {Object} page with fixed paddings
 */
const resolvePagePaddings = page => {
  const container = page.style;

  const style = evolve(
    {
      paddingTop: resolvePageVerticalPadding(container),
      paddingLeft: resolvePageHorizontalPadding(container),
      paddingRight: resolvePageHorizontalPadding(container),
      paddingBottom: resolvePageVerticalPadding(container),
    },
    page.style,
  );

  return Object.assign({}, page, { style });
};

/**
 * Translates all pages percentage paddings in fixed ones
 * This has to be computed from pages calculated size and not by Yoga
 * because at this point we didn't performed pagination yet.
 *
 * @param {Object} document root
 * @returns {Object} document root with translated page paddings
 */
const resolvePagesPaddings = root => {
  if (!root.children) return root;

  const children = root.children.map(resolvePagePaddings);

  return Object.assign({}, root, { children });
};

export default resolvePagesPaddings;
