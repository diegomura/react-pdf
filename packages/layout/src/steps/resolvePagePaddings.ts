import { evolve, matchPercent } from '@react-pdf/fns';
import { SafeStyle } from '@react-pdf/stylesheet';

import { SafeDocumentNode, SafePageNode } from '../types';

/**
 * Translates page percentage horizontal paddings in fixed ones
 *
 * @param container - Page container
 * @returns Resolve page horizontal padding
 */
const resolvePageHorizontalPadding =
  (container: SafeStyle) => (value: number) => {
    const match = matchPercent(value);
    const width = container.width as number;
    return match ? match.percent * width : value;
  };

/**
 * Translates page percentage vertical paddings in fixed ones
 *
 * @param container - Page container
 * @returns Resolve page vertical padding
 */
const resolvePageVerticalPadding =
  (container: SafeStyle) => (value: number) => {
    const match = matchPercent(value);
    const height = container.height as number;
    return match ? match.percent * height : value;
  };

/**
 * Translates page percentage paddings in fixed ones
 *
 * @param page
 * @returns Page with fixed paddings
 */
const resolvePagePaddings = (page: SafePageNode): SafePageNode => {
  const container = page.style;

  const style = evolve(
    {
      paddingTop: resolvePageVerticalPadding(container),
      paddingLeft: resolvePageHorizontalPadding(container),
      paddingRight: resolvePageHorizontalPadding(container),
      paddingBottom: resolvePageVerticalPadding(container),
    },
    page.style,
  );

  return Object.assign({}, page, { style });
};

/**
 * Translates all pages percentage paddings in fixed ones
 * This has to be computed from pages calculated size and not by Yoga
 * because at this point we didn't performed pagination yet.
 *
 * @param root - Document root
 * @returns Document root with translated page paddings
 */
const resolvePagesPaddings = (root: SafeDocumentNode) => {
  if (!root.children) return root;

  const children = root.children.map(resolvePagePaddings);

  return Object.assign({}, root, { children });
};

export default resolvePagesPaddings;
