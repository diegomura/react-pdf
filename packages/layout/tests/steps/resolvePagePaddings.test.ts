import { describe, expect, test } from 'vitest';

import resolvePagePaddings from '../../src/steps/resolvePagePaddings';

describe('layout resolvePagePaddings', () => {
  test('Should keep other styles untouched', () => {
    const result = resolvePagePaddings({
      type: 'DOCUMENT',
      props: {},
      children: [{ type: 'PAGE', props: {}, style: { color: 'red' } }],
    });

    expect(result).toMatchSnapshot();
  });

  test('Should leave numeric paddingTop as it is', () => {
    const result = resolvePagePaddings({
      type: 'DOCUMENT',
      props: {},
      children: [{ type: 'PAGE', props: {}, style: { paddingTop: 50 } }],
    });

    expect(result).toMatchSnapshot();
  });

  test('Should leave numeric paddingRight as it is', () => {
    const result = resolvePagePaddings({
      type: 'DOCUMENT',
      props: {},
      children: [{ type: 'PAGE', props: {}, style: { paddingRight: 50 } }],
    });

    expect(result).toMatchSnapshot();
  });

  test('Should leave numeric paddingBottom as it is', () => {
    const result = resolvePagePaddings({
      type: 'DOCUMENT',
      props: {},
      children: [{ type: 'PAGE', props: {}, style: { paddingBottom: 50 } }],
    });

    expect(result).toMatchSnapshot();
  });

  test('Should leave numeric paddingLeft as it is', () => {
    const result = resolvePagePaddings({
      type: 'DOCUMENT',
      props: {},
      children: [{ type: 'PAGE', props: {}, style: { paddingLeft: 50 } }],
    });

    expect(result).toMatchSnapshot();
  });

  test('Should resolve percent paddingTop', () => {
    const result = resolvePagePaddings({
      type: 'DOCUMENT',
      props: {},
      children: [
        {
          type: 'PAGE',
          props: {},
          style: { paddingTop: '10%' as any, width: 100, height: 200 },
        },
      ],
    });

    expect(result).toMatchSnapshot();
  });

  test('Should resolve percent paddingRight', () => {
    const result = resolvePagePaddings({
      type: 'DOCUMENT',
      props: {},
      children: [
        {
          type: 'PAGE',
          props: {},
          style: { paddingRight: '10%' as any, width: 100, height: 200 },
        },
      ],
    });

    expect(result).toMatchSnapshot();
  });

  test('Should resolve percent paddingBottom', () => {
    const result = resolvePagePaddings({
      type: 'DOCUMENT',
      props: {},
      children: [
        {
          type: 'PAGE',
          props: {},
          style: { paddingBottom: '10%' as any, width: 100, height: 200 },
        },
      ],
    });

    expect(result).toMatchSnapshot();
  });

  test('Should resolve percent paddingLeft', () => {
    const result = resolvePagePaddings({
      type: 'DOCUMENT',
      props: {},
      children: [
        {
          type: 'PAGE',
          props: {},
          style: { paddingLeft: '10%' as any, width: 100, height: 200 },
        },
      ],
    });

    expect(result).toMatchSnapshot();
  });

  test('Should resolve several pages', () => {
    const result = resolvePagePaddings({
      type: 'DOCUMENT',
      props: {},
      children: [
        {
          type: 'PAGE',
          props: {},
          style: { paddingTop: 10, width: 100, height: 200 },
        },
        {
          type: 'PAGE',
          props: {},
          style: { paddingBottom: '10%' as any, width: 100, height: 200 },
        },
        {
          type: 'PAGE',
          props: {},
          style: {
            paddingRight: 10,
            paddingLeft: '10%' as any,
            width: 100,
            height: 200,
          },
        },
      ],
    });

    expect(result).toMatchSnapshot();
  });
});
