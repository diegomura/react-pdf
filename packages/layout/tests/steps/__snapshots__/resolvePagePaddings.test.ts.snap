// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`layout resolvePagePaddings > Should keep other styles untouched 1`] = `
{
  "children": [
    {
      "props": {},
      "style": {
        "color": "red",
      },
      "type": "PAGE",
    },
  ],
  "props": {},
  "type": "DOCUMENT",
}
`;

exports[`layout resolvePagePaddings > Should leave numeric paddingBottom as it is 1`] = `
{
  "children": [
    {
      "props": {},
      "style": {
        "paddingBottom": 50,
      },
      "type": "PAGE",
    },
  ],
  "props": {},
  "type": "DOCUMENT",
}
`;

exports[`layout resolvePagePaddings > Should leave numeric paddingLeft as it is 1`] = `
{
  "children": [
    {
      "props": {},
      "style": {
        "paddingLeft": 50,
      },
      "type": "PAGE",
    },
  ],
  "props": {},
  "type": "DOCUMENT",
}
`;

exports[`layout resolvePagePaddings > Should leave numeric paddingRight as it is 1`] = `
{
  "children": [
    {
      "props": {},
      "style": {
        "paddingRight": 50,
      },
      "type": "PAGE",
    },
  ],
  "props": {},
  "type": "DOCUMENT",
}
`;

exports[`layout resolvePagePaddings > Should leave numeric paddingTop as it is 1`] = `
{
  "children": [
    {
      "props": {},
      "style": {
        "paddingTop": 50,
      },
      "type": "PAGE",
    },
  ],
  "props": {},
  "type": "DOCUMENT",
}
`;

exports[`layout resolvePagePaddings > Should resolve percent paddingBottom 1`] = `
{
  "children": [
    {
      "props": {},
      "style": {
        "height": 200,
        "paddingBottom": 20,
        "width": 100,
      },
      "type": "PAGE",
    },
  ],
  "props": {},
  "type": "DOCUMENT",
}
`;

exports[`layout resolvePagePaddings > Should resolve percent paddingLeft 1`] = `
{
  "children": [
    {
      "props": {},
      "style": {
        "height": 200,
        "paddingLeft": 10,
        "width": 100,
      },
      "type": "PAGE",
    },
  ],
  "props": {},
  "type": "DOCUMENT",
}
`;

exports[`layout resolvePagePaddings > Should resolve percent paddingRight 1`] = `
{
  "children": [
    {
      "props": {},
      "style": {
        "height": 200,
        "paddingRight": 10,
        "width": 100,
      },
      "type": "PAGE",
    },
  ],
  "props": {},
  "type": "DOCUMENT",
}
`;

exports[`layout resolvePagePaddings > Should resolve percent paddingTop 1`] = `
{
  "children": [
    {
      "props": {},
      "style": {
        "height": 200,
        "paddingTop": 20,
        "width": 100,
      },
      "type": "PAGE",
    },
  ],
  "props": {},
  "type": "DOCUMENT",
}
`;

exports[`layout resolvePagePaddings > Should resolve several pages 1`] = `
{
  "children": [
    {
      "props": {},
      "style": {
        "height": 200,
        "paddingTop": 10,
        "width": 100,
      },
      "type": "PAGE",
    },
    {
      "props": {},
      "style": {
        "height": 200,
        "paddingBottom": 20,
        "width": 100,
      },
      "type": "PAGE",
    },
    {
      "props": {},
      "style": {
        "height": 200,
        "paddingLeft": 10,
        "paddingRight": 10,
        "width": 100,
      },
      "type": "PAGE",
    },
  ],
  "props": {},
  "type": "DOCUMENT",
}
`;
